File: server/common/protocol.py
import socket
import struct

SEPARATOR = "|"
RECORD_SEP = "\n"
HEADER_SIZE = 4 

MSG_BET       = "BET"
MSG_ACK       = "ACK"
MSG_BATCH     = "BATCH"
MSG_BATCH_OK  = "BATCH_OK"
MSG_BATCH_ERR = "BATCH_ERR"
MSG_DONE      = "DONE"
MSG_WINNERS   = "WINNERS"


def _send_all(sock: socket.socket, data: bytes) -> None:
    """Send all bytes in data, avoiding short-writes."""
    sent = 0
    while sent < len(data):
        n = sock.send(data[sent:])
        if n == 0:
            raise OSError("Connection closed while sending data")
        sent += n


def _recv_all(sock: socket.socket, n: int) -> bytes:
    """Receive exactly n bytes, avoiding short-reads."""
    buf = b""
    while len(buf) < n:
        chunk = sock.recv(n - len(buf))
        if not chunk:
            raise OSError("Connection closed while receiving data")
        buf += chunk
    return buf


def send_message(sock: socket.socket, body: str) -> None:
    """Send a length-prefixed message."""
    data = body.encode("utf-8")
    _send_all(sock, struct.pack("!I", len(data)))
    _send_all(sock, data)


def recv_message(sock: socket.socket) -> str:
    """Receive a length-prefixed message."""
    length = struct.unpack("!I", _recv_all(sock, HEADER_SIZE))[0]
    return _recv_all(sock, length).decode("utf-8")


def recv_bet_fields(sock: socket.socket) -> tuple:
    """
    Receive a BET message and return:
    (agency, first_name, last_name, document, birthdate, number)
    """
    msg = recv_message(sock)
    parts = msg.split(SEPARATOR)
    if len(parts) != 7 or parts[0] != MSG_BET:
        raise ValueError(f"Invalid BET message: {msg!r}")
    _, agency, first_name, last_name, document, birthdate, number = parts
    return agency, first_name, last_name, document, birthdate, number


def send_ack(sock: socket.socket, document: str, number: str) -> None:
    send_message(sock, SEPARATOR.join([MSG_ACK, document, number]))


def recv_batch(sock: socket.socket) -> tuple[str, list[tuple]]:
    msg = recv_message(sock)
    lines = msg.split(RECORD_SEP)
    header_parts = lines[0].split(SEPARATOR)

    if header_parts[0] == MSG_DONE:
        if len(header_parts) != 2:
            raise ValueError(f"Invalid DONE message: {lines[0]!r}")
        return MSG_DONE, header_parts[1], []

    if header_parts[0] != MSG_BATCH or len(header_parts) != 3:
        raise ValueError(f"Invalid BATCH header: {lines[0]!r}")

    agency = header_parts[1]
    n_bets = int(header_parts[2])
    records = []
    for line in lines[1:]:
        if not line:
            continue
        fields = line.split(SEPARATOR)
        if len(fields) != 5:
            raise ValueError(f"Invalid bet record: {line!r}")
        records.append(tuple(fields))

    if len(records) != n_bets:
        raise ValueError(f"Expected {n_bets} bets, got {len(records)}")

    return MSG_BATCH, agency, records


def send_batch_ack(sock: socket.socket, cantidad: int, success: bool) -> None:
    msg_type = MSG_BATCH_OK if success else MSG_BATCH_ERR
    send_message(sock, SEPARATOR.join([msg_type, str(cantidad)]))


def send_done(sock: socket.socket, agency: str) -> None:
    send_message(sock, SEPARATOR.join([MSG_DONE, agency]))


def send_winners(sock: socket.socket, documents: list[str]) -> None:
    lines = [f"{MSG_WINNERS}{SEPARATOR}{len(documents)}"]
    lines.extend(documents)
    send_message(sock, RECORD_SEP.join(lines))


def recv_winners(sock: socket.socket) -> list[str]:
    msg = recv_message(sock)
    lines = msg.split(RECORD_SEP)
    header_parts = lines[0].split(SEPARATOR)
    if len(header_parts) != 2 or header_parts[0] != MSG_WINNERS:
        raise ValueError(f"Invalid WINNERS message: {lines[0]!r}")
    n = int(header_parts[1])
    documents = [l for l in lines[1:] if l]
    if len(documents) != n:
        raise ValueError(f"Expected {n} winners, got {len(documents)}")
    return documents

File: client/common/protocol.go
package common

import (
	"encoding/binary"
	"fmt"
	"net"
	"strconv"
	"strings"
)

const (
	msgTypeBet      = "BET"
	msgTypeAck      = "ACK"
	msgTypeBatch    = "BATCH"
	msgTypeBatchOK  = "BATCH_OK"
	msgTypeBatchErr = "BATCH_ERR"
	separator       = "|"
	recordSep       = "\n"
	headerSize      = 4
)

func sendAll(conn net.Conn, data []byte) error {
	sent := 0
	for sent < len(data) {
		n, err := conn.Write(data[sent:])
		if err != nil {
			return fmt.Errorf("sendAll: %w", err)
		}
		sent += n
	}
	return nil
}

func recvAll(conn net.Conn, n int) ([]byte, error) {
	buf := make([]byte, n)
	received := 0
	for received < n {
		r, err := conn.Read(buf[received:])
		if err != nil {
			return nil, fmt.Errorf("recvAll: %w", err)
		}
		received += r
	}
	return buf, nil
}

func sendMessage(conn net.Conn, body string) error {
	data := []byte(body)
	header := make([]byte, headerSize)
	binary.BigEndian.PutUint32(header, uint32(len(data)))
	
	if err := sendAll(conn, header); err != nil {
		return err
	}
	return sendAll(conn, data)
}

func recvMessage(conn net.Conn) (string, error) {
	header, err := recvAll(conn, headerSize)
	if err != nil {
		return "", err
	}
	length := binary.BigEndian.Uint32(header)
	body, err := recvAll(conn, int(length))
	if err != nil {
		return "", err
	}
	return string(body), nil
}

func SendBet(conn net.Conn, agency, firstName, lastName, document, birthdate, number string) error {
	fields := []string{msgTypeBet, agency, firstName, lastName, document, birthdate, number}
	return sendMessage(conn, strings.Join(fields, separator))
}

func RecvAck(conn net.Conn) (document, number string, err error) {
	msg, err := recvMessage(conn)
	if err != nil {
		return "", "", err
	}
	parts := strings.Split(msg, separator)
	if len(parts) != 3 || parts[0] != msgTypeAck {
		return "", "", fmt.Errorf("unexpected message: %s", msg)
	}
	return parts[1], parts[2], nil
}

type BetRecord struct {
	FirstName string
	LastName  string
	Document  string
	Birthdate string
	Number    string
}

func SendBatch(conn net.Conn, agency string, bets []BetRecord) error {
	var sb strings.Builder
	sb.WriteString(msgTypeBatch)
	sb.WriteString(separator)
	sb.WriteString(agency)
	sb.WriteString(separator)
	sb.WriteString(strconv.Itoa(len(bets)))
	for _, b := range bets {
		sb.WriteString(recordSep)
		sb.WriteString(strings.Join([]string{b.FirstName, b.LastName, b.Document, b.Birthdate, b.Number}, separator))
	}
	return sendMessage(conn, sb.String())
}

func RecvBatchAck(conn net.Conn) (cantidad int, success bool, err error) {
	msg, err := recvMessage(conn)
	if err != nil {
		return 0, false, err
	}
	parts := strings.Split(msg, separator)
	if len(parts) != 2 {
		return 0, false, fmt.Errorf("unexpected batch ack: %s", msg)
	}
	cantidad, err = strconv.Atoi(parts[1])
	if err != nil {
		return 0, false, fmt.Errorf("invalid cantidad in ack: %s", msg)
	}
	switch parts[0] {
	case msgTypeBatchOK:
		return cantidad, true, nil
	case msgTypeBatchErr:
		return cantidad, false, nil
	default:
		return 0, false, fmt.Errorf("unknown ack type: %s", parts[0])
	}
}